### PR TITLE
GitHub actions

### LINKED ISSUE #89: Adding GitHub Action for Jenkins Validation
See #88.

### PR DESCRIPTION
**Pull request by ash-sxn**
-  fixes #89 
- I still don't understand launching the job section completely  it was taken from Bruno's draft PR #88

### CODE REVIEW

File: .github/workflows/test-jenkins.yml [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            if [[ $BUILD_STATUS == "SUCCESS" ]]; then
+              echo "Job succeeded"
+            else
+              echo "Job failed"
+            fi
+          fi
+      - name: Teardown
+        if: env.NO_CHANGES != 'true'
+        run: |
+          ./jenkins_teardown.sh
```
Comment (berviantoleo) [inline .github/workflows/test-jenkins.yml:134]:
```suggestion
        run: ./jenkins_teardown.sh
```

File: .github/workflows/test-jenkins.yml [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+          # Installing dependencies 
+          # To check Sthe version of Jenkins, load the top page or any .../api/* page and check for the X-Jenkins response header. This contains the version number of Jenkins, like "1.404" This is also a good way to check if an URL is a Jenkins URL.
+          JENKINS_VERSION=$(curl -s -I -k http://admin:admin@127.0.0.1:8080 | grep -i '^X-Jenkins:' | awk '{print $2}')
+          echo "Jenkins version is: $JENKINS_VERSION"
+          # Before launching a job, we need to create a token for the admin user
+          CRUMB=$(curl -s -k http://admin:admin@127.0.0.1:8080/crumbIssuer/api/xml?xpath=concat\(//crumbRequestField,%22:%22,//crumb\) -c cookies.txt)
+          echo "CRUMB was found."
+          TOKEN=$(curl -s -k 'http://admin:admin@127.0.0.1:8080/user/admin/descriptorByName/jenkins.security.ApiTokenProperty/generateNewToken' --data 'newTokenName=kb-token' -b cookies.txt -H $CRUMB | jq -r '.data.tokenValue')
+          echo "TOKEN was found."
+          # Let's launch a job. The token has been generated in the Jenkins UI (admin, settings, and so on)
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:81]:
```suggestion
          # Let's launch a job. The token has been generated in the previous steps.
```

My bad. I wrote this comment before finding how to do it with the command line. 😊

File: .github/workflows/test-jenkins.yml [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            fi
+            sleep 5  # Adjust the sleep duration as needed
+          done
+          echo "This is BUILD__NUMBER $BUILD_NUMBER"
+          # Delay before retrieving build information
+          sleep 5
+          if [[ -z $BUILD_NUMBER || $BUILD_NUMBER == "null" ]]; then
+            echo "Job has never run"
+          else
+            echo "Job started. Build number: $BUILD_NUMBER"
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:109]:
```suggestion
          if [[ -z $BUILD_NUMBER || $BUILD_NUMBER == "null" ]]; then
          # If the build number is empty or "null", it means the job has never run
            echo "Job has never run"
          else
          # If the build number is not empty, the job has started and the build number is displayed
            echo "Job started. Build number: $BUILD_NUMBER"
          fi
```

File: .github/workflows/test-jenkins.yml [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+              BUILD_IN_PROGRESS=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080127.0.0.1:8080/job/$JOB_NAME_ENCODED/$BUILD_NUMBER/api/json | jq -r '.building')
+              echo "Build status: $BUILD_STATUS"
+              echo "Build in progress: $BUILD_IN_PROGRESS"
+
+              if [[ $BUILD_STATUS != "null" ]]; then
+                break
+              fi
+
+              sleep 5  # Adjust the sleep duration as needed
+            done
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:123]:
```suggestion
            # Wait for the job to complete
            echo "Waiting for the job to complete..."
            while true; do
            # Retrieve the build status and whether the build is in progress
              BUILD_STATUS=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080/job/$JOB_NAME_ENCODED/$BUILD_NUMBER/api/json | jq -r '.result')
              BUILD_IN_PROGRESS=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080127.0.0.1:8080/job/$JOB_NAME_ENCODED/$BUILD_NUMBER/api/json | jq -r '.building')
              echo "Build status: $BUILD_STATUS"
              echo "Build in progress: $BUILD_IN_PROGRESS"
              # If the build status is not "null", it means the build has been completed
              if [[ $BUILD_STATUS != "null" ]]; then
                break
              fi
              sleep 5  # Adjust the sleep duration as needed
            done
```

File: .github/workflows/test-jenkins.yml [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+          # If the message is found, awk exits with a non-zero status (1), and the loop continues.
+          # If the message is not found, the loop exits, and the "Jenkins is running" message is displayed.
+          timeout 60 bash -c 'until curl -s -f http://127.0.0.1:8080/login > /dev/null; do sleep 5; done' && echo "Jenkins is running" || echo "Jenkins is not running"
+          echo "Jenkins is ready"
+
+      - name: Run curl command to test the stack
+        run: |
+          set -x
+          # Installing dependencies 
+          # To check Sthe version of Jenkins, load the top page or any .../api/* page and check for the X-Jenkins response header. This contains the version number of Jenkins, like "1.404" This is also a good way to check if an URL is a Jenkins URL.
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:186]:
```suggestion
          # To check the version of Jenkins, load the top page or any .../api/* page and check for the X-Jenkins response header. This contains the version number of Jenkins, like "1.404" This is also a good way to check if an URL is a Jenkins URL.
```

File: .github/workflows/test-jenkins.yml [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+          JENKINS_VERSION=$(curl -s -I -k http://admin:admin@127.0.0.1:8080 | grep -i '^X-Jenkins:' | awk '{print $2}')
+          echo "Jenkins version is: $JENKINS_VERSION"
+          # Before launching a job, we need to create a token for the admin user
+          CRUMB=$(curl -s -k http://admin:admin@127.0.0.1:8080/crumbIssuer/api/xml?xpath=concat\(//crumbRequestField,%22:%22,//crumb\) -c cookies.txt)
+          echo "CRUMB was found."
+          TOKEN=$(curl -s -k 'http://admin:admin@127.0.0.1:8080/user/admin/descriptorByName/jenkins.security.ApiTokenProperty/generateNewToken' --data 'newTokenName=kb-token' -b cookies.txt -H $CRUMB | jq -r '.data.tokenValue')
+          echo "TOKEN was found."
+          # Let's launch a job. The token has been generated in the Jenkins UI (admin, settings, and so on)
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:194]:
```suggestion
          # Let's launch a job. The token has been generated in the previous steps.
```

File: .github/workflows/test-jenkins.yml [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+          echo "Launching a job"
+          JOB_NAME="%28simple%29%20demo%20job"
+          JOB_NAME_ENCODED=$(echo "$JOB_NAME" | awk '{ gsub(/ /, "%20"); gsub(/\(/, "%28"); gsub(/\)/, "%29"); print }')
+          echo "JOB_NAME_ENCODED is $JOB_NAME_ENCODED"
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:198]:
```suggestion
          echo "Launching a job"
          # Define the job name with parentheses
          JOB_NAME="%28simple%29%20demo%20job"
          # Encode the job name to handle special characters
          JOB_NAME_ENCODED=$(echo "$JOB_NAME" | awk '{ gsub(/ /, "%20"); gsub(/\(/, "%28"); gsub(/\)/, "%29"); print }')
          echo "JOB_NAME_ENCODED is $JOB_NAME_ENCODED"
```

File: .github/workflows/test-jenkins.yml [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+          # Will have to switch to "simple-java-maven-app" later on.
+          # For the time being, it does not work as we haven't yet customized the config.xml file with the end user's GitHub handle.
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:200]:
```suggestion
```

File: .github/workflows/test-jenkins.yml [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            echo "Retrieved build info: $BUILD_INFO"
+            BUILD_NUMBER=$(echo $BUILD_INFO | jq -r '.lastBuild.number')
+            BUILD_IN_PROGRESS=$(echo $BUILD_INFO | jq -r '.lastBuild.building')
+            echo "Build number: $BUILD_NUMBER"
+            echo "Build in progress: $BUILD_IN_PROGRESS"
+            if [[ -n $BUILD_NUMBER && $BUILD_IN_PROGRESS == "true" ]]; then
+              break
+            fi
+            sleep 5  # Adjust the sleep duration as needed
+          done
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:215]:
```suggestion
          # Wait for the job to start running
          echo "Waiting for the job to start running..."
          # Initialize the build number
          BUILD_NUMBER=""
          # Loop until the build number is retrieved
          while [[ -z $BUILD_NUMBER ]]; do
            # Retrieve build information for the job
            BUILD_INFO=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080/job/$JOB_NAME_ENCODED/api/json)
            echo "Retrieved build info: $BUILD_INFO"
            # Extract the build number and check if the job is in progress
            BUILD_NUMBER=$(echo $BUILD_INFO | jq -r '.lastBuild.number')
            BUILD_IN_PROGRESS=$(echo $BUILD_INFO | jq -r '.lastBuild.building')
            echo "Build number: $BUILD_NUMBER"
            echo "Build in progress: $BUILD_IN_PROGRESS"
            # Break the loop if the build number is retrieved and the job is in progress
            if [[ -n $BUILD_NUMBER && $BUILD_IN_PROGRESS == "true" ]]; then
              break
            fi
            sleep 5  # Adjust the sleep duration as needed
          done
```

File: .github/workflows/test-jenkins.yml [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+              BUILD_IN_PROGRESS=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080127.0.0.1:8080/job/$JOB_NAME_ENCODED/$BUILD_NUMBER/api/json | jq -r '.building')
+              echo "Build status: $BUILD_STATUS"
+              echo "Build in progress: $BUILD_IN_PROGRESS"
+  
+              if [[ $BUILD_STATUS = "null" ]]; then
+                break
+              fi
+  
+              sleep 5  # Adjust the sleep duration as needed
+            done
```
Comment (gounthar) [inline .github/workflows/test-jenkins.yml:237]:
```suggestion
            echo "Job started. Build number: $BUILD_NUMBER"
            # Display the message indicating that the job has started along with the build number
            # Wait for the job to complete
            echo "Waiting for the job to complete..."
            # Start an infinite loop to continuously check the job status
            while true; do
              # Retrieve the build status and job progress using API calls
              BUILD_STATUS=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080/job/$JOB_NAME_ENCODED/$BUILD_NUMBER/api/json | jq -r '.result')
              BUILD_IN_PROGRESS=$(curl -s -k http://admin:$TOKEN@127.0.0.1:8080127.0.0.1:8080/job/$JOB_NAME_ENCODED/$BUILD_NUMBER/api/json | jq -r '.building')
              # Display the build status and progress
              echo "Build status: $BUILD_STATUS"
              echo "Build in progress: $BUILD_IN_PROGRESS"
              # Break the loop if the build status is not available (null)
              if [[ $BUILD_STATUS = "null" ]]; then
                break
              fi
              sleep 5  # Adjust the sleep duration as needed
            done
```

**Review by gounthar — COMMENTED** [review]
A few things to change here and there.
I'm still searching for the documentation.

**Comment by gounthar** [comment]
@ash-sxn 
https://community.jenkins.io/t/can-we-create-a-user-token-through-the-api/8319/5?u=poddingue

File: .github/workflows/test-jenkins.yml
Comment (berviantoleo) [inline .github/workflows/test-jenkins.yml:1]: I have a general comment. How about making the steps by steps as some shell scripts?

For example:

```
      - name: running ${{ matrix.dir }}
        run: ./init.sh ${{ matrix.dir}}
            
      - name: checking files running the jenkins_init.sh with argument
        run: ./check.sh $ARG

```

I'm thinking of reusing some of the scripts for later on.

Comment (berviantoleo) [reply .github/workflows/test-jenkins.yml:1]: Anyway, we can address this in another separate task. It's a bit tricky when we move to the scripts.

Overall, I'm good with the current solutions.

**Review by gounthar — APPROVED** [review]

**Pull request merged**